Extensions/Lighting/lightobstacleruntimebehavior.js

```diff
@@ -0,0 +1,115 @@
+gdjs.LightObstaclesManager = function (runtimeScene, sharedData) {
+  this._obstacleRBush = new rbush(9, [
+    '.owner.getAABB().min[0]',
+    '.owner.getAABB().min[1]',
+    '.owner.getAABB().max[0]',
+    '.owner.getAABB().max[1]',
+  ]);
+};
+
+gdjs.LightObstaclesManager.getManager = function (runtimeScene) {
```

Comment:
@4ian I have a question/doubt. If getManager is on the prototype (ie, `gdjs.LightObstaclesManager.prototype.getManager`), I get errors saying getManager is not a function. Why is that so? Does it have something related to being a singleton (if I understand the term correctly, singleton means only one instance of the class right?), or is it something else?

Reply:
@HarsimranVirk When you put a method on the prototype in JavaScript, it's akin to a "class method", meaning that you have to instantiate an instance of the class:
```
MyClass = function() { this.something = 3; } // This is the "constructor"
MyClass.prototype.myMethod = function() { return this.something; } // This is a "class method"

var instance = new MyClass();
console.log(instance.myMethod()); // Will display 3.
MyClass.myMethod() // error because myMethod does not exist.
```

if you remove the `.prototype` when declaring a function on a class, you're just setting a function as a property... in other words it's a "static" function:

```
MyClass = function() { ... } // This is the "constructor"
MyClass.myStaticMethod = function() { return "Hello world";  } // This is a "static" method.

console.log(MyClass.myStaticMethod()); // Will display "Hello world".
```

Here, `getManager` is a "static" function. You can think of it as a helper. It was put here because it's better to put it as a property of `gdjs.LightObstaclesManager` rather than as a global variable. This is indeed more or less related to a "singleton" pattern, though a bit different. A singleton is a *single* class instance for the whole code, while here we want *one manager per scene*.
How do we do that? By having a single function called `gdjs.LightObstaclesManager.getManager`, that you can call anywhere, and that returns the gdjs.LightObstaclesManager for the scene. The way it does it is by creating (if necessary) a `new gdjs.LightObstaclesManager()`, storing it inside the runtimeScene (but we could imagine using a hashtable or something like that to avoid polluting the runtimeScene... but this is the fastest way to do it), and then return it.

Note that `getManager` is a static method so you must always call it by writing the entire name `var lightObstaclesManager = gdjs.LightObstaclesManager.getManager(runtimeScene);`.
You can then call usual (non static) methods:
`lightObstaclesManager.getObstaclesAroundObject(object, radius, ...);` etc

Reply:
Note that in recent versions of JavaScript, the `class` keyword was added, so you can do instead:

```js
class MyClass {
  constructor() {
    this.something = 3;
  }
  
  myMethod() { return this.something; } 
  static myStaticMethod() { return "hello world"; }
}


var instance = new MyClass();
console.log(instance.myMethod()); // Will display 3.
console.log(MyClass.myStaticMethod()); // Will display "Hello world".
```

but this is just "syntactic sugar" over the `prototype` that I explained before. We say that JavaScript is a "prototype" based language. In GDJS, we don't use the class keyword because it's unsupported in old browsers, notably Android 4.4 (we write "ES5" Javascript, not ES6).

Reply:
> we write "ES5" Javascript, not ES6

I totally forgot that 😅 I've been using a lot of `const`, `let` and arrow functions, might need some refactoring then (especially with using `this` in `function()`).

Reply:
That's ok as a first draft, just don't forget to make a card about this in the Trello so we remember to make a pass to ensure everything is using `var`/`function` :)